#!/usr/bin/env python3
"""
–¢–µ—Å—Ç –ø–æ–¥–∫–ª—é—á–µ–Ω–∏—è –∫ production —Å–µ—Ä–≤–µ—Ä—É —á–µ—Ä–µ–∑ –∫–ª–∏–µ–Ω—Ç—Å–∫–∏–π –º–æ–¥—É–ª—å
"""
import sys
import asyncio
import logging
from pathlib import Path

# –î–æ–±–∞–≤–ª—è–µ–º –ø—É—Ç–∏
project_root = Path(__file__).parent
sys.path.insert(0, str(project_root))
sys.path.insert(0, str(project_root / "client"))

# –ù–∞—Å—Ç—Ä–æ–π–∫–∞ –ª–æ–≥–∏—Ä–æ–≤–∞–Ω–∏—è
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

async def test_production_connection():
    """–¢–µ—Å—Ç –ø–æ–¥–∫–ª—é—á–µ–Ω–∏—è –∫ production —Å–µ—Ä–≤–µ—Ä—É"""
    try:
        from modules.grpc_client.core.grpc_client import GrpcClient
        
        logger.info("=" * 60)
        logger.info("–¢–µ—Å—Ç –ø–æ–¥–∫–ª—é—á–µ–Ω–∏—è –∫ production —Å–µ—Ä–≤–µ—Ä—É")
        logger.info("=" * 60)
        
        # –°–æ–∑–¥–∞–µ–º –∫–ª–∏–µ–Ω—Ç —Å –∫–æ–Ω—Ñ–∏–≥—É—Ä–∞—Ü–∏–µ–π –¥–ª—è production
        config = {
            'servers': {
                'production': {
                    'address': '20.151.51.172',
                    'port': 443,
                    'use_ssl': True,
                    'ssl_verify': False,  # –í—Ä–µ–º–µ–Ω–Ω–æ –æ—Ç–∫–ª—é—á–µ–Ω–æ –¥–ª—è self-signed
                    'use_http2': True,
                    'timeout': 300,
                    'retry_attempts': 1,  # –û–¥–Ω–∞ –ø–æ–ø—ã—Ç–∫–∞ –¥–ª—è –±—ã—Å—Ç—Ä–æ–≥–æ —Ç–µ—Å—Ç–∞
                    'retry_delay': 1.0,
                    'keepalive': True
                }
            },
            'auto_fallback': False,
            'connection_timeout': 60,
            'max_retry_attempts': 1
        }
        
        client = GrpcClient(config)
        
        logger.info(f"üîå –ü–æ–ø—ã—Ç–∫–∞ –ø–æ–¥–∫–ª—é—á–µ–Ω–∏—è –∫ production —Å–µ—Ä–≤–µ—Ä—É...")
        logger.info(f"   –ê–¥—Ä–µ—Å: 20.151.51.172:443")
        logger.info(f"   SSL: True")
        logger.info(f"   –¢–∞–π–º–∞—É—Ç: 60 —Å–µ–∫—É–Ω–¥")
        
        # –ü–µ—Ä–µ–∫–ª—é—á–∞–µ–º—Å—è –Ω–∞ production —Å–µ—Ä–≤–µ—Ä –∏ –ø–æ–¥–∫–ª—é—á–∞–µ–º—Å—è
        await client.switch_server('production')
        connected = await client.connect('production')
        
        if connected:
            logger.info("‚úÖ –ü–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –∫ production —Å–µ—Ä–≤–µ—Ä—É —É—Å–ø–µ—à–Ω–æ!")
            
            # –ü—Ä–æ–±—É–µ–º –ø—Ä–æ—Å—Ç–æ–π RPC –≤—ã–∑–æ–≤
            try:
                logger.info("üß™ –¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ InterruptSession...")
                from modules.grpc_client.proto import streaming_pb2
                
                request = streaming_pb2.InterruptRequest(
                    hardware_id="test_connection"
                )
                
                # –ü–æ–ª—É—á–∞–µ–º stub –∏–∑ connection_manager
                if not hasattr(client.connection_manager, 'stub') or not client.connection_manager.stub:
                    logger.error("‚ùå Stub –Ω–µ–¥–æ—Å—Ç—É–ø–µ–Ω")
                    return False
                
                stub = client.connection_manager.stub
                
                # –í—ã–∑—ã–≤–∞–µ–º RPC –Ω–∞–ø—Ä—è–º—É—é
                result = await stub.InterruptSession(request, timeout=10.0)
                
                if result and result.success:
                    logger.info("‚úÖ InterruptSession —É—Å–ø–µ—à–µ–Ω!")
                else:
                    logger.warning("‚ö†Ô∏è InterruptSession –≤–µ—Ä–Ω—É–ª –Ω–µ–æ–∂–∏–¥–∞–Ω–Ω—ã–π —Ä–µ–∑—É–ª—å—Ç–∞—Ç")
                    
            except Exception as e:
                logger.error(f"‚ùå –û—à–∏–±–∫–∞ –ø—Ä–∏ —Ç–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–∏ RPC: {e}")
            
            # –û—Ç–∫–ª—é—á–∞–µ–º—Å—è
            await client.disconnect()
            logger.info("üîå –û—Ç–∫–ª—é—á–µ–Ω–æ –æ—Ç —Å–µ—Ä–≤–µ—Ä–∞")
            
            return True
        else:
            logger.error("‚ùå –ù–µ —É–¥–∞–ª–æ—Å—å –ø–æ–¥–∫–ª—é—á–∏—Ç—å—Å—è –∫ production —Å–µ—Ä–≤–µ—Ä—É")
            return False
            
    except Exception as e:
        logger.error(f"‚ùå –û—à–∏–±–∫–∞ –ø—Ä–∏ —Ç–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–∏: {e}")
        import traceback
        logger.error(traceback.format_exc())
        return False

async def main():
    """–ì–ª–∞–≤–Ω–∞—è —Ñ—É–Ω–∫—Ü–∏—è"""
    success = await test_production_connection()
    return 0 if success else 1

if __name__ == "__main__":
    exit_code = asyncio.run(main())
    sys.exit(exit_code)
